File: src/statemotion/progress/progressEngine.ts
// StateMotion — Phase 0.1 vertical slice: progress engine (TypeScript).
//
// Host-independent preview-side mirror of the C++ engine (ticket 015: separate
// implementations, shared reviewed fixtures, parity within 1e-6). Reuses the
// generated enum values from parameterIds.ts — no duplicated numbers.
//
// Knows nothing about Premiere / UXP / Adobe APIs.

import { ProgressMode, AlignmentMode } from '../../../shared/generated/parameterIds.ts';
import { EasingMode, type EasingCurve, evaluateEasing } from './easing.ts';

export type ProgressErrorCode = 'None' | 'NonFiniteInput';

export interface ProgressInput {
  visibleElapsedSeconds: number;
  visibleDurationSeconds: number;
  transitionDurationSeconds: number;
  delaySeconds: number;
  alignment: AlignmentMode;
  mode: ProgressMode;
  manualProgress: number;
  // Easing configuration. Named modes ignore `curve`; only CUSTOM uses it.
  easing: EasingMode;
  curve: EasingCurve;
}

export interface ProgressResult {
  linearProgress: number;
  easedProgress: number;
}

export interface ProgressOutput {
  ok: boolean;
  error: ProgressErrorCode;
  result: ProgressResult;
}

function clampLower0(v: number): number {
  return v < 0 ? 0 : v;
}

function finite(v: number): boolean {
  return Number.isFinite(v);
}

export function computeLinearProgress(
  alignment: AlignmentMode,
  elapsed: number,
  visibleDuration: number,
  transitionDuration: number,
  delay: number,
): number {
  elapsed = clampLower0(elapsed);
  delay = clampLower0(delay);
  transitionDuration = clampLower0(transitionDuration);

  if (alignment === AlignmentMode.EntireClip) {
    if (visibleDuration <= 0) return 1;
    return Math.min(1, Math.max(0, elapsed / visibleDuration));
  }
  if (alignment === AlignmentMode.ClipStart) {
    if (visibleDuration <= 0) return 0;
    const ws = Math.min(Math.max(delay, 0), visibleDuration);
    const we = Math.min(Math.max(delay + transitionDuration, 0), visibleDuration);
    if (we <= ws) return elapsed >= ws ? 1 : 0;
    return Math.min(1, Math.max(0, (elapsed - ws) / (we - ws)));
  }
  // ClipEnd
  if (visibleDuration <= 0) return 1;
  const we = Math.min(Math.max(visibleDuration - delay, 0), visibleDuration);
  const ws = Math.min(Math.max(visibleDuration - delay - transitionDuration, 0), visibleDuration);
  if (we <= ws) return elapsed >= ws ? 1 : 0;
  return Math.min(1, Math.max(0, (elapsed - ws) / (we - ws)));
}

export function evaluateProgress(input: ProgressInput): ProgressOutput {
  if (
    !finite(input.visibleElapsedSeconds) ||
    !finite(input.visibleDurationSeconds) ||
    !finite(input.transitionDurationSeconds) ||
    !finite(input.delaySeconds) ||
    !finite(input.manualProgress)
  ) {
    return { ok: false, error: 'NonFiniteInput', result: { linearProgress: 0, easedProgress: 0 } };
  }

  const easing = input.easing ?? EasingMode.EASE_IN_OUT;
  const curve = input.curve ?? { x1: 0.42, y1: 0.0, x2: 0.58, y2: 1.0 };

  let linear: number;
  if (input.mode === ProgressMode.Manual) {
    linear = Math.min(1, Math.max(0, input.manualProgress));
  } else {
    linear = computeLinearProgress(
      input.alignment,
      input.visibleElapsedSeconds,
      input.visibleDurationSeconds,
      input.transitionDurationSeconds,
      input.delaySeconds,
    );
  }

  let eased = 0;
  switch (input.mode) {
    case ProgressMode.HoldA:
      eased = 0; // exact endpoint, never eased
      break;
    case ProgressMode.HoldB:
      eased = 1; // exact endpoint, never eased
      break;
    case ProgressMode.Manual:
      // Manual mode bypasses automatic easing: manualProgress is the direct
      // normalized progress between A and B (deterministic midpoint).
      eased = linear;
      break;
    case ProgressMode.AToB:
      eased = evaluateEasing(easing, curve, linear);
      break;
    case ProgressMode.BToA:
      eased = 1 - evaluateEasing(easing, curve, linear);
      break;
    case ProgressMode.AToBToA:
      eased = linear <= 0.5
        ? evaluateEasing(easing, curve, 2 * linear)
        : 1 - evaluateEasing(easing, curve, 2 * linear - 1);
      break;
    case ProgressMode.BToAToB:
      eased = linear <= 0.5
        ? 1 - evaluateEasing(easing, curve, 2 * linear)
        : evaluateEasing(easing, curve, 2 * linear - 1);
      break;
  }
  return { ok: true, error: 'None', result: { linearProgress: linear, easedProgress: eased } };
}

File: src/statemotion/progress/progress_engine.cpp
// StateMotion — Phase 0.1 vertical slice: progress engine implementation.
// See progress_engine.h for scope, clean-room note, and invalid-input rule.

#include "progress_engine.h"

#include <cmath>

namespace statemotion {

namespace {

bool finite(double v) { return std::isfinite(v); }

// Documented input rule (ticket 013 + handoff):
//  - Non-finite (NaN / +-Inf) on any timing/manual field -> rejected (typed error).
//  - Negative elapsed/delay/transitionDuration are normalized by clamping to 0.
//    Negative visibleDuration is handled per-alignment (rest state, see below).
double clampLower0(double v) { return v < 0.0 ? 0.0 : v; }

}  // namespace

double computeLinearProgress(ids::AlignmentMode alignment, double elapsed, double visibleDuration,
                             double transitionDuration, double delay) {
    elapsed = clampLower0(elapsed);
    delay = clampLower0(delay);
    transitionDuration = clampLower0(transitionDuration);

    if (alignment == ids::AlignmentMode::EntireClip) {
        // Ignores transitionDuration and delay. Zero/invalid visible duration is
        // deterministic destination (ticket 013: "ENTIRE_CLIP zero/invalid duration
        // -> deterministic destination").
        if (visibleDuration <= 0.0) return 1.0;
        return std::clamp(elapsed / visibleDuration, 0.0, 1.0);
    }

    // CLIP_START: window starts at `delay`. CLIP_END: window ends at visibleDuration-delay.
    if (alignment == ids::AlignmentMode::ClipStart) {
        if (visibleDuration <= 0.0) return 0.0;  // no timeline -> rest at start (A)
        const double ws = std::clamp(delay, 0.0, visibleDuration);
        const double we = std::clamp(delay + transitionDuration, 0.0, visibleDuration);
        if (we <= ws) return (elapsed >= ws) ? 1.0 : 0.0;  // instantaneous step
        return std::clamp((elapsed - ws) / (we - ws), 0.0, 1.0);
    }

    // ClipEnd
    if (visibleDuration <= 0.0) return 1.0;  // no timeline -> rest at destination (B)
    const double we = std::clamp(visibleDuration - delay, 0.0, visibleDuration);
    const double ws = std::clamp(visibleDuration - delay - transitionDuration, 0.0, visibleDuration);
    if (we <= ws) return (elapsed >= ws) ? 1.0 : 0.0;  // instantaneous step
    return std::clamp((elapsed - ws) / (we - ws), 0.0, 1.0);
}

ProgressOutput evaluateProgress(const ProgressInput& input) {
    ProgressOutput out;
    if (!finite(input.visibleElapsedSeconds) || !finite(input.visibleDurationSeconds) ||
        !finite(input.transitionDurationSeconds) || !finite(input.delaySeconds) ||
        !finite(input.manualProgress)) {
        out.ok = false;
        out.error = ProgressErrorCode::NonFiniteInput;
        return out;
    }

    double linear;
    if (input.mode == ids::ProgressMode::Manual) {
        linear = std::clamp(input.manualProgress, 0.0, 1.0);  // ignores alignment timing
    } else {
        linear = computeLinearProgress(input.alignment, input.visibleElapsedSeconds,
                                       input.visibleDurationSeconds, input.transitionDurationSeconds,
                                       input.delaySeconds);
    }
    out.result.linearProgress = linear;

    double eased = 0.0;
    switch (input.mode) {
        case ids::ProgressMode::HoldA:
            eased = 0.0;
            break;
        case ids::ProgressMode::HoldB:
            eased = 1.0;
            break;
        case ids::ProgressMode::Manual:
            eased = smoothstep(linear);
            break;
        case ids::ProgressMode::AToB:
            eased = smoothstep(linear);
            break;
        case ids::ProgressMode::BToA:
            eased = 1.0 - smoothstep(linear);
            break;
        case ids::ProgressMode::AToBToA:  // two independently-eased halves, never one triangle
            eased = (linear <= 0.5) ? smoothstep(2.0 * linear)
                                    : 1.0 - smoothstep(2.0 * linear - 1.0);
            break;
        case ids::ProgressMode::BToAToB:
            eased = (linear <= 0.5) ? 1.0 - smoothstep(2.0 * linear)
                                    : smoothstep(2.0 * linear - 1.0);
            break;
    }
    out.result.easedProgress = eased;
    return out;
}

}  // namespace statemotion
